File: NERP.NetworkEntities/SpliceClosureService.cs
using Common.Config;
using Common.DataAccess.Models;
using Common.DataAccess.Mongo;
using Microsoft.Extensions.Options;
using NERP.NetworkEntities.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;

namespace NERP.NetworkEntities
{
    public class SpliceClosureService : ISpliceClosureService
    {
        private readonly IRepository<SpliceClosure> _spliceClosuresRepository;
        private const string SpliceClosuresRepository = "MST_SpliceClosures";

        public SpliceClosureService(IOptions<DbConfig> dbConfig)
        {
            var databaseConfig = dbConfig.Value;

            var dbSettings = new DbSettings { ConnectionString = databaseConfig.ConnectionString, Database = databaseConfig.Database };

            _spliceClosuresRepository = new MongoRepository<SpliceClosure>(dbSettings, SpliceClosuresRepository);
        }

        public async Task<SpliceClosure> GetSpliceClosure(string spliceClosureId)
        {
            if (string.IsNullOrEmpty(spliceClosureId)) return null;

            return await _spliceClosuresRepository.GetByIdAsync(spliceClosureId);
        }

        public async Task<IEnumerable<SpliceClosure>> GetSpliceClosures(SearchFilter searchFilter = null, string orderBy = null, SortOrder sortOrder = SortOrder.Descending, int? page = null, int? pageSize = null)
        {
            Expression<Func<SpliceClosure, bool>> filterExpression = x => true;

            if (searchFilter != null && searchFilter.Filters.Count > 0)
                filterExpression = ExpressionBuilder.GetExpression<SpliceClosure>(searchFilter);

            IList<SpliceClosure> spliceClosures = null;

            if (page != null && pageSize != null)
                spliceClosures = (await _spliceClosuresRepository.GetPaginatedAsync(filterExpression, orderBy, sortOrder, (int)page, (int)pageSize)).ToList();
            else
                spliceClosures = (await _spliceClosuresRepository.GetAllAsync(filterExpression, null, orderBy, sortOrder)).ToList();

            return spliceClosures;
        }

        public async Task<long> GetSpliceClosuresCount(SearchFilter searchFilter = null)
        {
            Expression<Func<SpliceClosure, bool>> filterExpression = x => true;

            if (searchFilter != null && searchFilter.Filters.Count > 0)
            {
                filterExpression = ExpressionBuilder.GetExpression<SpliceClosure>(searchFilter);
            }

            return await _spliceClosuresRepository.CountAsync(filterExpression);
        }

        public async Task<string> CheckNameExists(string name)
        {
            var spliceClosure = await _spliceClosuresRepository.GetOneAsync(x => x.Name == name);

            return spliceClosure?.Id;
        }

        public async Task<SpliceClosure> SaveSpliceClosure(SpliceClosure spliceClosure)
        {
            return await _spliceClosuresRepository.UpdateOneAsync(spliceClosure);
        }

        public async Task<bool> DeleteSpliceClosure(string spliceClosureId)
        {
            return (await _spliceClosuresRepository.DeleteByIdAsync(spliceClosureId)) == 1;
        }
    }
}

File: NERP.RND.Web/ClientApp/src/app/entities/services/spliceClosure.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient, HttpParams } from "@angular/common/http";

import { ServiceResponse } from '../../common/models/serviceResponse';
import { SpliceClosure } from '../models/spliceClosure';
import { PaginationRequest, SortOrder } from '../../common/models/paginationRequest';

@Injectable()
export class SpliceClosureService {

    getSpliceClosuresApi = 'SpliceClosures/GetSpliceClosures/';
    getSpliceClosureApi = 'SpliceClosures/GetSpliceClosure/';
    saveSpliceClosureApi = 'SpliceClosures/SaveSpliceClosure/';
    deleteSpliceClosureApi = 'SpliceClosures/DeleteSpliceClosure/';

    constructor(private http: HttpClient) { }

    getSpliceClosures(paginationRequest?: PaginationRequest): Observable<ServiceResponse> {

        if (paginationRequest == null)
            paginationRequest = {
                page: null,
                pageSize: null,
                filter: null,
                order: {
                    orderByProperty: "updatedDate",
                    sortOrder: SortOrder.descending
                }
            };

        return this.http.post<ServiceResponse>(this.getSpliceClosuresApi, paginationRequest);
    }

    getSpliceClosure(spliceClosureId): Observable<ServiceResponse> {

        const params = new HttpParams()
            .set('spliceClosureId', spliceClosureId);

        return this.http.get<ServiceResponse>(this.getSpliceClosureApi, { params });
    }

    saveSpliceClosure(spliceClosure: SpliceClosure): Observable<ServiceResponse> {
        return this.http.post<ServiceResponse>(this.saveSpliceClosureApi, spliceClosure);
    }

    deleteSpliceClosure(spliceClosureId: string): Observable<ServiceResponse> {

        const params = new HttpParams()
            .set('spliceClosureId', spliceClosureId);

        return this.http.delete<ServiceResponse>(this.deleteSpliceClosureApi, { params });
    }
}
